File: leetcode/path_sum/PathSum.cpp
#include <iostream> 

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:    
    bool hasPathSum(TreeNode *root, int sum) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        if (!root) {
            return false; 
        }
        
        if (root->left == NULL && root->right == NULL) {
            if (root->val == sum) {
                return true; 
            }
            else {
                return false; 
            }
        }
        
        int remain = sum - root->val; 
                
        return hasPathSum(root->left, remain) || hasPathSum(root->right, remain);
    }
};

int main() {
    Solution sln; 
    int sum = 6; 
    TreeNode * root = new TreeNode(1); 
    root->left = new TreeNode(2); 
    root->left->left = new TreeNode(3); 
    root->left->left->left = new TreeNode(4); 
    root->left->left->left->left = new TreeNode(5); 
    
    if (sln.hasPathSum(root, sum)) {
        cout << "true" << endl; 
    }
    else {
        cout << "false" << endl; 
    }
    return 0; 
}

File: leetcode6/path_sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 * 
 * Check whether a binary tree has a root-to-leaf path sum that is equal 
 * to the given target. 
 * Solve it recursively. The idea was to compare the remaining sum with 
 * the value of current node. If current node is a leaf and its value is 
 * equal to the remaining sum, then we found a path. Otherwise, recursively
 * move onto left and right child with remaining target value. 
 */
public class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) {
            return false; 
        }
        
        if (root.left == null && root.right == null && root.val == sum) {
            return true; 
        }
        
        boolean result = hasPathSum(root.left, sum - root.val); 
        
        if (result) {
            return true; 
        }
        
        return hasPathSum(root.right, sum - root.val); 
    }
}
